// pdf-creator-node/dist/pdf/puppeteer.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.htmlPdfTokensToPuppeteer = htmlPdfTokensToPuppeteer;
exports.renderPdfToBuffer = renderPdfToBuffer;
exports.writePdfToFile = writePdfToFile;
exports.bufferToStream = bufferToStream;
const promises_1 = require("fs/promises");
const path_1 = require("path");
const puppeteer_1 = __importDefault(require("puppeteer"));
const stream_1 = require("stream");
/** Maps html-pdf `{{page}}` / `{{pages}}` to Puppeteer header/footer template tokens. */
function htmlPdfTokensToPuppeteer(html) {
    return html
        .replace(/\{\{\s*page\s*\}\}/gi, '<span class="pageNumber"></span>')
        .replace(/\{\{\s*pages\s*\}\}/gi, '<span class="totalPages"></span>');
}
function wrapPrintTemplate(html) {
    return `<div style="font-size:10px;width:100%;box-sizing:border-box;padding:4px 12px;">${htmlPdfTokensToPuppeteer(html)}</div>`;
}
function borderToMargin(border) {
    var _a, _b, _c, _d;
    if (border == null)
        return undefined;
    if (typeof border === "string") {
        return { top: border, right: border, bottom: border, left: border };
    }
    return {
        top: (_a = border.top) !== null && _a !== void 0 ? _a : "0",
        right: (_b = border.right) !== null && _b !== void 0 ? _b : "0",
        bottom: (_c = border.bottom) !== null && _c !== void 0 ? _c : "0",
        left: (_d = border.left) !== null && _d !== void 0 ? _d : "0",
    };
}
function ensureFullHtmlDocument(html) {
    const trimmed = html.trim();
    if (/^<!DOCTYPE/i.test(trimmed) || /<html[\s>]/i.test(trimmed)) {
        return html;
    }
    return `<!DOCTYPE html><html><head><meta charset="utf-8"></head><body>${html}</body></html>`;
}
/**
 * v4+: Puppeteer only supports one footer template — we use `default`, else `first`, else `last`.
 * Per-page numeric keys are ignored.
 */
function pickFooterHtml(footer) {
    const c = footer === null || footer === void 0 ? void 0 : footer.contents;
    if (!c)
        return undefined;
    if (c.default != null && String(c.default).trim() !== "")
        return String(c.default);
    if (c.first != null && String(c.first).trim() !== "")
        return String(c.first);
    if (c.last != null && String(c.last).trim() !== "")
        return String(c.last);
    return undefined;
}
function withBrowser(fn) {
    return __awaiter(this, void 0, void 0, function* () {
        const browser = yield puppeteer_1.default.launch({
            headless: true,
            args: ["--no-sandbox", "--disable-setuid-sandbox", "--disable-dev-shm-usage"],
        });
        try {
            return yield fn(browser);
        }
        finally {
            yield browser.close();
        }
    });
}
function buildPdfOptions(opts) {
    var _a, _b;
    const margin = borderToMargin(opts.border);
    const headerHtml = (_a = opts.header) === null || _a === void 0 ? void 0 : _a.contents;
    const footerHtml = pickFooterHtml(opts.footer);
    const hasHeader = Boolean(headerHtml && headerHtml.trim());
    const hasFooter = Boolean(footerHtml && footerHtml.trim());
    const format = opts.width && opts.height ? undefined : (_b = opts.format) !== null && _b !== void 0 ? _b : "A4";
    const landscape = opts.orientation === "landscape";
    const pdfOpts = {
        printBackground: true,
        landscape,
    };
    if (hasHeader || hasFooter) {
        pdfOpts.displayHeaderFooter = true;
        pdfOpts.headerTemplate = hasHeader
            ? wrapPrintTemplate(headerHtml)
            : "<span></span>";
        pdfOpts.footerTemplate = hasFooter
            ? wrapPrintTemplate(footerHtml)
            : "<span></span>";
    }
    if (margin) {
        pdfOpts.margin = margin;
    }
    if (opts.width && opts.height) {
        pdfOpts.width = opts.width;
        pdfOpts.height = opts.height;
    }
    else if (format) {
        pdfOpts.format = format;
    }
    return pdfOpts;
}
/**
 * Renders HTML to PDF bytes using headless Chromium (Puppeteer).
 */
function renderPdfToBuffer(html, options) {
    return __awaiter(this, void 0, void 0, function* () {
        var _a;
        const fullHtml = ensureFullHtmlDocument(html);
        const pdfOpts = buildPdfOptions(options);
        const timeout = (_a = options.timeout) !== null && _a !== void 0 ? _a : 30000;
        return withBrowser((browser) => __awaiter(this, void 0, void 0, function* () {
            var _a;
            const page = yield browser.newPage();
            page.setDefaultTimeout(timeout);
            if (options.httpHeaders && Object.keys(options.httpHeaders).length > 0) {
                yield page.setExtraHTTPHeaders(options.httpHeaders);
            }
            if ((_a = options.httpCookies) === null || _a === void 0 ? void 0 : _a.length) {
                yield page.setCookie(...options.httpCookies);
            }
            yield page.setContent(fullHtml, Object.assign({ waitUntil: "load", timeout }, (options.base ? { baseURL: options.base } : {})));
            if (typeof options.renderDelay === "number" && options.renderDelay > 0) {
                yield new Promise((r) => setTimeout(r, options.renderDelay));
            }
            const buf = yield page.pdf(pdfOpts);
            return Buffer.from(buf);
        }));
    });
}
function writePdfToFile(path, buffer) {
    return __awaiter(this, void 0, void 0, function* () {
        yield (0, promises_1.mkdir)((0, path_1.dirname)(path), { recursive: true });
        yield (0, promises_1.writeFile)(path, buffer);
        return { filename: path };
    });
}
function bufferToStream(buffer) {
    return stream_1.Readable.from(buffer);
}
